### PR TITLE
[WC, PT] Store compression scale in f16

### PR DESCRIPTION
**Pull request by nikita-savelyevv**
### Changes

- Store compression scale if FP16
- Add type conversion to original data type after decompression

Below are the compression subgraphs for the first conv2d in mobilenet_v2 after conversion to OV, this is similar to the table presented in #2537 .
image
Compared to OV case, there is an additional Multiply node after the scale Multiply node. It seems to come from Batch Norm applied to the convolution. In case of PT weight compression it does not get merged into the weight as it does in OV case.

### Reason for changes

Weight compression for PT backend fails when applied to model in half precision. The reason is that the scale is always in FP32, and hence decompression result is also in FP32, which conflicts with input type of FP16.

### Related tickets

134063

### Tests

Added test for half/full precision cases. Also added cases for different devices as it was thought that it may influence tracing in half precision.

### CODE REVIEW

File: nncf/torch/quantization/layers.py [hunk 1]
```diff
@@ -1039,14 +1039,18 @@ class WeightsDecompressor(nn.Module):
     Applies decompression of compressed weights in the forward pass
     """
 
-    def __init__(self, scale: torch.Tensor, zero_point: torch.Tensor):
+    def __init__(self, scale: torch.Tensor, zero_point: torch.Tensor, result_dtype: torch.dtype = None):
```
Comment (alexsu52) [inline nncf/torch/quantization/layers.py:1042]: Could you explain why `result_dtype` is optional?

Comment (nikita-savelyevv) [reply nncf/torch/quantization/layers.py:1042]: The idea was to also support the case when no explicit casting is needed

**Review by alexsu52 — APPROVED** [review]
LGTM
